### PR TITLE
Plugins session invalidate

### PR DESCRIPTION
**Pull request by vladak**
fixes bunch of problems in `AuthorizationFramework`.

Notable changes:
  - HTTP session invalidation now happens in `AuthorizationFramework` as opposed to in the plugins
  - random string generation moved from plugin tests to `src/opengrok` so that all tests can use it
  - some bad object programming practices softened

### CODE REVIEW

File: src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java [hunk 1]
```diff
@@ -141,6 +151,38 @@ public void setPluginDirectory(String directory) {
...
+     */
+    public void setLoadClasses(boolean flag) {
+        loadClasses = flag;
+    }
+    
+    /**
+     * Whether to search for plugins in class files.
+     * @return true if enabled, false otherwise
+     */
+    public boolean getLoadClasses() {
```
Comment (tulinkry) [inline src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java]: better naming for boolean property getter?

File: src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java [hunk 2]
```diff
@@ -141,6 +151,38 @@ public void setPluginDirectory(String directory) {
...
+     */
+    public void setLoadJars(boolean flag) {
+        loadJars = flag;
+    }
+    
+    /**
+     * Whether to search for plugins in class files.
+     * @return true if enabled, false otherwise
+     */
+    public boolean getLoadJars() {
```
Comment (tulinkry) [inline src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java]: better naming for boolean property getter?

File: src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java [hunk 3]
```diff
@@ -683,31 +762,52 @@ private boolean checkAll(HttpServletRequest request, String cache, Nameable enti
...
+        
+        lock.readLock().lock();
+        try {
+            // Make sure there is a HTTP session that corresponds to current plugin version.
+            HttpSession session;
+            if (((session = request.getSession(false)) != null) && isSessionInvalid(session)) {
+                session.invalidate();
+                stats.addRequest("authorization_sessions_invalidated");
+            }
+            request.getSession().setAttribute(SESSION_VERSION, getPluginVersion());
```
Comment (tulinkry) [inline src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java]: does this need to be locked under the lock if this does not interfere with the stack?

File: src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java [hunk 4]
```diff
@@ -141,6 +151,38 @@ public void setPluginDirectory(String directory) {
...
+    }
+    
+    /**
+     * Whether to search for plugins in class files.
+     * @return true if enabled, false otherwise
+     */
+    public boolean getLoadJars() {
+        return loadJars;
+    }
+
```
Comment (tulinkry) [inline src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java]: and also `loadPluginsFiles` or simiar for both of these may be of use in the future

File: src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java [hunk 5]
```diff
@@ -543,16 +605,16 @@ public Object run() {
             }
         });
 
-        // clone a new stack not interfering with the current stack
-        AuthorizationStack newStack = RuntimeEnvironment.getInstance().getPluginStack().clone();
-
-        // increase the current plugin version tracked by the framework
-        increasePluginVersion();
+        // Clone a new stack not interfering with the current stack.
+        AuthorizationStack newStack = getStack().clone();
```
Comment (tulinkry) [inline src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java]: Only thing here is that maybe I overlooked where will be the new stack comming from the configuration change?

File: src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java [hunk 6]
```diff
@@ -683,31 +762,52 @@ private boolean checkAll(HttpServletRequest request, String cache, Nameable enti
...
-
-        boolean overallDecision = performCheck(entity, pluginPredicate, skippingPredicate);
+        long time = 0;
+        boolean overallDecision = false;
+        
+        lock.readLock().lock();
+        try {
+            // Make sure there is a HTTP session that corresponds to current plugin version.
+            HttpSession session;
+            if (((session = request.getSession(false)) != null) && isSessionInvalid(session)) {
```
Comment (tulinkry) [inline src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java]: Can you please check if this works with the JSONServlet?

From now we kinda assume that plugins use http session. But I guess this is the common use case.

File: src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java (hunk 1 above)
Comment (vladak) [reply src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java]: fixed

File: src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java (hunk 2 above)
Comment (vladak) [reply src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java]: fixed

File: src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java (hunk 4 above)
Comment (vladak) [reply src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java]: if you mean a setter for both I am hesitant to add unused code.

File: src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java (hunk 5 above)
Comment (vladak) [reply src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java]: `RuntimeEnvironment#applyConfig()` now contains this:

```
         // set the new plugin directory and reload the authorization framework
         getAuthorizationFramework().setPluginDirectory(config.getPluginDirectory());
+        getAuthorizationFramework().setStack(config.getPluginStack());
         getAuthorizationFramework().reload();
```

File: src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java (hunk 4 above)
Comment (tulinkry) [reply src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java]: ok

File: src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java (hunk 5 above)
Comment (tulinkry) [reply src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java]: but now what happens if a request comes between these two lines
```java
         getAuthorizationFramework().setStack(config.getPluginStack());
         // <-- request for authorization
         getAuthorizationFramework().reload();
```

Comment (vladak) [reply src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java]: good point. I guess delayed reconfiguration can be used to address that.

File: src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java [hunk 7]
```diff
@@ -74,6 +74,10 @@
      * Stack of available plugins/stacks in the order of the execution.
      */
     AuthorizationStack stack;
+    /**
+     * New stack.
```
Comment (tulinkry) [inline src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java]: add a brief purpose info

File: src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java [hunk 8]
```diff
@@ -605,19 +605,22 @@ public Object run() {
...
         });
 
-        // Clone a new stack not interfering with the current stack.
-        AuthorizationStack newStack = getStack().clone();
-
+        AuthorizationStack newLocalStack = this.newStack;
+        if (newLocalStack == null) {
+            // Clone a new stack not interfering with the current stack.
+            newLocalStack = getStack().clone();
+        }
```
Comment (tulinkry) [inline src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java]: I'd like to see also `this.newStack.clone()` so that noone from the outside can modify the stack.

File: src/org/opensolaris/opengrok/web/JSONSearchServlet.java [hunk 9]
```diff
@@ -109,6 +111,13 @@ protected void doGet(HttpServletRequest req, HttpServletResponse resp)
             result.put(PARAM_HIST, hist);
         }
 
+        if (type != null) {
+            type = URLDecoder.decode(type);
+            engine.setType(type);
+            valid = true;
+            result.put(PARAM_TYPE, type);
+        }
+
```
Comment (tulinkry) [inline src/org/opensolaris/opengrok/web/JSONSearchServlet.java]: how is this relevant?

Comment (tulinkry) [reply src/org/opensolaris/opengrok/web/JSONSearchServlet.java]: oh I see the issue

File: src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java (hunk 6 above)
Comment (vladak) [reply src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java]: Checked the servlet with `http://localhost:8084/source/json?freetext=foo&maxresults=80`, returns valid JSON.

File: src/org/opensolaris/opengrok/web/JSONSearchServlet.java (hunk 9 above)
Comment (vladak) [reply src/org/opensolaris/opengrok/web/JSONSearchServlet.java]: yep, just modified it while I was there.

File: src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java (hunk 3 above)
Comment (tulinkry) [reply src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java]: ok, clear enough

File: src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java (hunk 6 above)
Comment (tulinkry) [reply src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java]: perfect

File: src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java (hunk 7 above)
Comment (vladak) [reply src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java]: added

File: src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java (hunk 8 above)
Comment (vladak) [reply src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java]: fixed

**Comment by tarzanek** [comment]
merging ;)
